File: mizugos/procs/define.go
package procs

// Processor 處理器介面
//
// 負責將網路傳輸的封包與程式內部的訊息物件之間做轉換, 並在收到訊息後呼叫對應的處理函式(Process)
//
// 職責:
//   - Encode: 將內部訊息物件轉換成封包格式(序列化)
//   - Decode: 將封包格式轉換回內部訊息物件(反序列化)
//   - Process: 接收封包後, 呼叫對應的訊息處理函式
//   - Add / Del / Get: 管理訊息處理函式的註冊與查詢
//
// 建立自訂處理器的流程:
//  1. 定義訊息結構, 必須包含 messageID 欄位 (int32), 作為訊息識別
//  2. 若使用 Protobuf, 可將定義檔放在 support/proto-mizugo 目錄下, 由編譯器自動生成程式碼
//  3. 定義一個結構實作 Processor 介面, 並內含 Procmgr, 只需自行實作 Encode / Decode / Process 三個函式, 其他訊息處理管理功能由 Procmgr 提供
//  4. 將自訂 Processor 加入到框架中使用
//
// mizugo 內建的處理器:
//   - Json: 使用 JSON 格式序列化/反序列化訊息
//   - Proto: 使用 Protobuf 格式序列化/反序列化訊息
//   - Raven: 使用 Raven 格式序列化/反序列化訊息
type Processor interface {
	// Encode 封包編碼, 將內部訊息物件轉換成封包資料, 用於傳輸
	Encode(input any) (output any, err error)

	// Decode 封包解碼, 將封包資料轉換回內部訊息物件, 用於後續處理
	Decode(input any) (output any, err error)

	// Process 訊息處理, 當解碼成功後, 會呼叫對應的訊息處理函式
	Process(input any) error

	// Add 新增訊息處理
	Add(messageID int32, process Process)

	// Del 刪除訊息處理
	Del(messageID int32)

	// Get 取得訊息處理
	Get(messageID int32) Process
}

// Process 訊息處理函式類型
//   - 每一個 messageID 對應一個 Process 函式
//   - 當訊息解碼成功後, 會呼叫對應的 Process 並傳入訊息物件
//   - 使用者需自行將 message 轉換成正確型別 (例如 *msgs.Json / *msgs.Proto)
type Process func(message any)

File: support/client-unity/Packages/com.fouridstudio.mizugo-client-unity/Runtime/Define.cs
namespace Mizugo
{
    /// <summary>
    /// 事件編號
    /// </summary>
    public enum EventID
    {
        /// <summary>
        /// 無效事件
        /// </summary>
        Unknow = 0,

        /// <summary>
        /// 連線事件, 參數是null
        /// </summary>
        Connect,

        /// <summary>
        /// 斷線事件, 參數是null
        /// </summary>
        Disconnect,

        /// <summary>
        /// 接收事件, 參數是null
        /// </summary>
        Recv,

        /// <summary>
        /// 傳送事件, 參數是null
        /// </summary>
        Send,

        /// <summary>
        /// 錯誤事件, 參數是Exception
        /// </summary>
        Error,

        /// <summary>
        /// 訊息事件, 參數是訊息物件; 外部無法使用
        /// </summary>
        Message,
    }

    /// <summary>
    /// 網路定義
    /// </summary>
    public class Define
    {
        /// <summary>
        /// 標頭長度
        /// </summary>
        public const int headerSize = 2;

        /// <summary>
        /// 封包長度
        /// </summary>
        public const int packetSize = ushort.MaxValue;
    }

    /// <summary>
    /// 客戶端介面
    /// </summary>
    public interface IClient
    {
        /// <summary>
        /// 連線處理
        /// </summary>
        public void Connect(string host, int port);

        /// <summary>
        /// 斷線處理
        /// </summary>
        public void Disconnect();

        /// <summary>
        /// 更新處理
        /// </summary>
        public void Update();

        /// <summary>
        /// 傳送訊息
        /// </summary>
        public void Send(object message);

        /// <summary>
        /// 設定事件處理器
        /// </summary>
        public void SetEvent(IEventmgr eventmgr);

        /// <summary>
        /// 新增事件處理
        /// </summary>
        public void AddEvent(EventID eventID, OnTrigger onEvent);

        /// <summary>
        /// 刪除事件處理
        /// </summary>
        public void DelEvent(EventID eventID);

        /// <summary>
        /// 設定訊息處理器
        /// </summary>
        public void SetProc(IProcmgr procmgr);

        /// <summary>
        /// 新增訊息處理
        /// </summary>
        public void AddProcess(int messageID, OnTrigger onProcess);

        /// <summary>
        /// 刪除訊息處理
        /// </summary>
        public void DelProcess(int messageID);

        /// <summary>
        /// 設定編碼/解碼
        /// </summary>
        public void SetCodec(params ICodec[] codec);

        /// <summary>
        /// 設定標頭長度, 連線後就無法變更
        /// </summary>
        public void SetHeaderSize(int size);

        /// <summary>
        /// 設定封包長度, 連線後就無法變更
        /// </summary>
        public void SetPacketSize(int size);

        /// <summary>
        /// 取得是否連線
        /// </summary>
        public bool IsConnect { get; }

        /// <summary>
        /// 取得是否需要處理事件, 連線中或是事件佇列中有事件都會需要處理事件
        /// </summary>
        public bool IsUpdate { get; }
    }

    /// <summary>
    /// 事件處理器介面
    /// </summary>
    public interface IEventmgr
    {
        /// <summary>
        /// 事件處理
        /// </summary>
        public void Process(EventID eventID, object param);

        /// <summary>
        /// 新增事件處理
        /// </summary>
        public void Add(EventID eventID, OnTrigger onEvent);

        /// <summary>
        /// 刪除事件處理
        /// </summary>
        public void Del(EventID eventID);
    }

    /// <summary>
    /// 訊息處理器介面
    /// </summary>
    public interface IProcmgr
    {
        /// <summary>
        /// 訊息處理
        /// </summary>
        public void Process(object input);

        /// <summary>
        /// 新增訊息處理
        /// </summary>
        public void Add(int messageID, OnTrigger onProcess);

        /// <summary>
        /// 刪除訊息處理
        /// </summary>
        public void Del(int messageID);

        /// <summary>
        /// 取得訊息處理
        /// </summary>
        public OnTrigger Get(int messageID);
    }

    /// <summary>
    /// 編碼/解碼介面
    /// </summary>
    public interface ICodec
    {
        /// <summary>
        /// 編碼處理
        /// </summary>
        public object Encode(object input);

        /// <summary>
        /// 解碼處理
        /// </summary>
        public object Decode(object input);
    }

    /// <summary>
    /// 處理函式類型
    /// </summary>
    public delegate void OnTrigger(object param);
}
